### PR TITLE
[Persistence] Refactors BlockStore Interface

### PR DESCRIPTION
**Pull request by dylanlott**
## Description

This PR refactors the persistence module to create the BlockStore interface. In doing so, it stops higher level components from directly calling the KVStore and exposes a specialized GetBlock and StoreBlock function instead. These two will eventually be wrapped with savepoints and rollbacks logic to facilitate atomic commits to the block store.

## Issue

Parts of #562

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Creates the BlockStore interface and replaces mocks to use it where possible
- Forces height and block serialization handling down into the BlockStore module, making it cleaner and easier to get any block from the bus.

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: persistence/blockstore/block_store.go [hunk 1]
```diff
@@ -6,13 +6,19 @@ import (
...
 
 // BlockStore is a key-value store that maps block heights to serialized
 // block structures.
 // * It manages the atomic state transitions for applying a Unit of Work.
 type BlockStore interface {
-	kvstore.KVStore
+	kvstore.KVStore // TODO_IN_THIS_COMMIT remove this eventually and expose only the two below
+
+	GetBlock(height uint64) (*coreTypes.Block, error)
+	StoreBlock(height uint64, block *coreTypes.Block) error
```
Comment (h5law) [inline persistence/blockstore/block_store.go:21]: 🔥 So much cleaner this way

**Review by h5law — CHANGES_REQUESTED** [review]
Great job so far, sure you know already but there are a few places in the rest of the codebase that need to be updated (`rpc/handlers_query.go` not sure of any others off the top of my head) but personally, I don't see anything stopping this from getting into the wild as soon as you're done. ❤️

File: consensus/e2e_tests/utils_test.go [hunk 2]
```diff
@@ -375,37 +374,73 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel, bus modules.Bus)
...
-			},
-		}
-		return codec.GetCodec().Marshal(blockWithHeight)
-	}).AnyTimes()
+	blockStoreMock.
+		EXPECT().
+		StoreBlock(gomock.Any(), gomock.Any()).
+		DoAndReturn(func(height []byte, block *coreTypes.Block) error {
+			return nil
+		}).
```
Comment (Olshansk) [inline consensus/e2e_tests/utils_test.go:382]: Optional NIT: if there's nothing you actually want to do, I think you might be able to get away with just `Return(nil)`. Not 100% sure thought so you'd need to double check.

File: consensus/e2e_tests/utils_test.go [hunk 3]
```diff
@@ -375,37 +374,73 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel, bus modules.Bus)
...
+		AnyTimes()
 
-	persistenceMock.EXPECT().GetBlockStore().Return(blockStoreMock).AnyTimes()
+	blockStoreMock.
```
Comment (Olshansk) [inline consensus/e2e_tests/utils_test.go:385]:
```suggestion
	// NB: The business logic in this mock and below is vital for testing state-sync end-to-end
	blockStoreMock.
```

File: consensus/state_sync/server.go [hunk 4]
```diff
@@ -72,8 +69,10 @@ func (m *stateSync) HandleGetBlockRequest(blockReq *typesCons.GetBlockRequest) e
 	}
 
 	// get block from the persistence module
-	block, err := m.getBlockAtHeight(blockReq.Height)
+	blockStore := m.GetBus().GetPersistenceModule().GetBlockStore()
+	block, err := blockStore.GetBlock(blockReq.Height)
 	if err != nil {
+		m.logger.Error().Err(typesCons.ErrConsensusMempoolFull).Msg(typesCons.DisregardHotstuffMessage)
```
Comment (Olshansk) [inline consensus/state_sync/server.go:75]: Why are we logging a `ErrConsensusMempoolFull` error here?

File: persistence/blockstore/block_store.go [hunk 5]
```diff
@@ -43,6 +49,31 @@ func NewBlockStore(path string) (BlockStore, error) {
 	}, nil
 }
 
+// StoreBlock takes a coreType Block and stores it.
```
Comment (Olshansk) [inline persistence/blockstore/block_store.go:52]:
```suggestion
// StoreBlock accepts a coreType Block and stores it for the given height.
```

File: persistence/blockstore/block_store.go [hunk 6]
```diff
@@ -43,6 +49,31 @@ func NewBlockStore(path string) (BlockStore, error) {
...
+		return err
+	}
+	// TECHDEBT add a proper logger to blockstore
+	// bs.logger.Info().Uint64("height", block.BlockHeader.Height).Msg("Storing block in block store")
+	return bs.kv.Set(utils.HeightToBytes(height), b)
+}
+
+// GetBlock returns a coreTypes Block at the given height.
+func (bs *blockStore) GetBlock(height uint64) (*coreTypes.Block, error) {
+	var block coreTypes.Block
```
Comment (Olshansk) [inline persistence/blockstore/block_store.go:65]: NIT: Either move this above the line where we first use it or use named return types.

File: persistence/blockstore/block_store.go [hunk 7]
```diff
@@ -43,6 +49,31 @@ func NewBlockStore(path string) (BlockStore, error) {
...
+	blockBytes, err := bs.kv.Get(utils.HeightToBytes(height))
+	if err != nil {
+		return nil, err
+	}
+	err = codec.GetCodec().Unmarshal(blockBytes, &block)
+	if err != nil {
```
Comment (Olshansk) [inline persistence/blockstore/block_store.go:71]: NIT: Single line

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
> rpc/handlers_query.go

@dylanlott If you end up updating `rpc/handlers_query.go` per @h5law's comment, could you also add an E2E test? Doesn't need to be exhaustive but figured it might be a good time to just add one in w/e endpoint we're modifying.

File: consensus/state_sync/server.go (hunk 4 above)
Comment (dylanlott) [reply consensus/state_sync/server.go:75]: Good catch, this was an artifact of the refactor that I was apparently context-blind to. Updated to return a better error. 👍

**Review by Olshansk — APPROVED** [review]

**Comment by dylanlott** [comment]
> > rpc/handlers_query.go
> 
> @dylanlott If you end up updating `rpc/handlers_query.go` per @h5law's comment, could you also add an E2E test? Doesn't need to be exhaustive but figured it might be a good time to just add one in w/e endpoint we're modifying.

Addressed in #786

**Pull request merged**